File: algo/ShellSort.java
public class ShellSort {

    int sort(int arr[]) {
        int n = arr.length;
    
        for (int gap = n/2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i += 1) {
                int temp = arr[i];
                int j;
                for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                    arr[j] = arr[j - gap];
                arr[j] = temp;
            }
        }
        return 0;
    }
    public static void main(String[] args) {
        ShellSort ob = new ShellSort();
        int arr[] = {12, 34, 54, 2, 3};
        ob.sort(arr);
        System.out.println("Sorted array");
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }
}

File: algo/ShellSort.js

function shellSort(arr) {
    let n = arr.length;
    for (let gap = Math.floor(n/2); gap > 0; gap = Math.floor(gap/2)) {
        for (let i = gap; i < n; i += 1)  {
            let temp = arr[i];

            let j;
            for (j = i; j >= gap && arr[j-gap] > temp; j-=gap)  {
                arr[j] = arr[j-gap];
            }
            arr[j] = temp;
        }
    }
    return arr;
}

arr = [12, 34, 54, 2, 3];
console.log("Array before sorting: " + arr);
console.log("Array after sorting: " + shellSort(arr));
